### DISCUSSION TITLE
Pushing messages to GCP Pub/Sub

### DISCUSSION THREAD

**Discussion by BostjanBozic** [vernemq/vernemq#2073] [Q&A]
Hello,

I have a question regarding some integration of VerneMQ. My idea would be to use VerneMQ as MQTT broker, but then push all messages to GCP Pub/Sub, so that I can easily integrate data streams with different GCP services.
I checked around the issues and I found one where Pub/Sub is mentioned, but no concrete information is given (https://github.com/vernemq/vernemq/issues/705).

Do you know, is that somehow possible?

Thank you and best regards,
Bostjan

**Comment by ioolkos** [comment]
@BostjanBozic I suppose this would have to be enabled in a VerneMQ plugin.
I'm not too familiar with GCP PubSub. We'd have to determine what messaging protocol this implements and how to forward messages.

Happy to collect more info on this topic here.

**Comment by BostjanBozic** [comment]
Hello,

thank you for feedback. I did some digging around, playing with possibility of creating an Erlang plugin based on [example](https://github.com/redclawtech/vernemq_kinesis). This is this example is using another AWS API for Erlang, which includes all the logic on how to interact with AWS Kinesis. Unfortunately I did not find anything like that when it comes to GCP, meaning that most likely this would have to be written as well.

For messaging protocol, to my understanding `protobuf` is used.
When it comes to pushing the messages, I think data itself can be of any type since it gets serialized to a byte stream using `protobuf`.

**Comment by ioolkos** [comment]
I see. Yeah, I'm not aware of a library like `erlcloud` (which implements AWS APIs) for GCP.
What the plugin maybe could do is push messages over HTTP; I haven't consulted the GCP PubSub docs on that, though.
